File: problem_10.js
var primes = function(n) {

  var nums = [null, null],
      primes = [];

  // 0 and 1 aren't prime but we want the indexes to match up.
  for (var i = 2; i < n; i++) {
    nums.push(true);
  }

  // Go up to the sqrt of the given number, as per the formula.
  for (var i = 2; i <= Math.sqrt(n); i++) {
    if (nums[i]) {
      // Start with the square of i and step with i each time to get all multiples.
      for (var j = i * i; j < n; j += i) {
        // Set multiples to false as they aren't primes.
        nums[j] = false;
      }
    }
  }

  // All of the remaining elements that are 'true' have a prime value for their index.
  for (var i = 2; i < n; i++) {
    if (nums[i]) {
      primes.push(i);
    }
  }

  console.log(primes.reduce(function(a, b) { return a + b; }));
}

primes(2000000);

File: problem_10.rb
class Sieve
  def initialize(n)
    @n = n
  end

  def primes
    nums = [nil, nil, *2..@n]

    # For every number from 2 to the sqrt of the given number...
    (2..Math.sqrt(@n)).each do |i|
      # ...square it and step through its multiples up to n.
      # Set it to nil if it's in the range we're going through.
      (i**2..@n).step(i){|m| nums[m] = nil}  if nums[i]
      # 'if nums[i]' is checking if the current number in the range is present in the range of
      # 2..Math.sqrt(@n). If it is then we need to set it to nil, as it's not prime.
    end

    p nums.compact.inject(:+)
  end
end


Sieve.new(2_000_000).primes